### ISSUE TITLE
zipfile build fails in circleci (docker) context

### ISSUE THREAD

**Issue by mikemccabe** [aws-cloudformation/cloudformation-cli-python-plugin#95]
`cfn submit --dry` works fine locally, but fails when run in circleci.

TL;DR this is nested docker, I suspect that sharing files twice with -v is failing.  My guess: The inner -v still uses the docker daemon's view of the filesystem, and paths aren't shared correctly.  Switch to using the docker Volume API?  https://docs.docker.com/engine/reference/commandline/volume_create/

ADDED from comment below, suggested fix in https://circleci.com/docs/2.0/building-docker-images/#mounting-folders

Discussion:

```cfn submit --dry || (echo 'cat rpdk.log' && cat rpdk.log && false)
Starting Docker build. This may take several minutes if the image 'lambci/lambda:build-python3.7' needs to be pulled first.
=== Caught downstream error ===
Command 'pip install --no-cache-dir --no-color --disable-pip-version-check --upgrade --find-links /project --requirement /project/requirements.txt --target /project/build' in image 'lambci/lambda:build-python3.7' returned non-zero exit status 1
---
If debugging indicates this is a possible error with this program,
please report the issue to the team and include the log file 'rpdk.log'.
Issue tracker: https://github.com/aws-cloudformation/aws-cloudformation-rpdk/issues
cat RPDK.log
[2020-05-03T19:13:44Z] DEBUG    - Logging set up successfully
[2020-05-03T19:13:44Z] DEBUG    - Running submit: Namespace(command=<function submit at 0x7fc7af35b200>, dry_run=True, endpoint_url=None, region=None, role_arn=None, set_default=False, subparser_name='submit', use_role=True, verbose=0, version=False)
[2020-05-03T19:13:44Z] DEBUG    - Root directory: /home/circleci/my-app/sendgrid-whitelabel-domain
[2020-05-03T19:13:44Z] DEBUG    - Loading project file '/home/circleci/my-app/sendgrid-whitelabel-domain/.rpdk-config'
[2020-05-03T19:13:44Z] INFO     - Validating your resource specification...
[2020-05-03T19:13:44Z] DEBUG    - Resolved '#/definitions/DnsEntry' to 'file:///home/circleci/my-app/sendgrid-whitelabel-domain/sendgrid-whitelabel-domain.json#/definitions/DnsEntry'
[2020-05-03T19:13:44Z] DEBUG    - Parsed 'file:///home/circleci/my-app/sendgrid-whitelabel-domain/sendgrid-whitelabel-domain.json#/definitions/DnsEntry' to '(<BASE>, 'definitions', 'DnsEntry')'
[2020-05-03T19:13:44Z] DEBUG    - Edge from '(<BASE>, 'properties', 'Dns', 'items')' to '(<BASE>, 'definitions', 'DnsEntry')'
[2020-05-03T19:13:44Z] DEBUG    - Rewriting refs in '<BASE>' (file:///home/circleci/my-app/sendgrid-whitelabel-domain/sendgrid-whitelabel-domain.json)
[2020-05-03T19:13:44Z] DEBUG    -   '#/definitions/DnsEntry' -> '#/definitions/DnsEntry'
[2020-05-03T19:13:44Z] DEBUG    - overrides.json not found. Not writing to package.
[2020-05-03T19:13:44Z] DEBUG    - Package started
[2020-05-03T19:13:44Z] DEBUG    - '/home/circleci/my-app/sendgrid-whitelabel-domain/build' not found, skipping removal
Traceback (most recent call last):
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/python/codegen.py", line 192, in _remove_build_artifacts
    shutil.rmtree(deps_path)
  File "/home/circleci/.pyenv/versions/3.7.7/lib/python3.7/shutil.py", line 485, in rmtree
    onerror(os.lstat, path, sys.exc_info())
  File "/home/circleci/.pyenv/versions/3.7.7/lib/python3.7/shutil.py", line 483, in rmtree
    orig_st = os.lstat(path)
FileNotFoundError: [Errno 2] No such file or directory: '/home/circleci/my-app/sendgrid-whitelabel-domain/build'
[2020-05-03T19:13:44Z] DEBUG    - Dependencies build started from '/home/circleci/my-app/sendgrid-whitelabel-domain'
[2020-05-03T19:13:44Z] DEBUG    - command is 'pip install --no-cache-dir --no-color --disable-pip-version-check --upgrade --find-links /project --requirement /project/requirements.txt --target /project/build'
[2020-05-03T19:13:44Z] WARNING  - Starting Docker build. This may take several minutes if the image 'lambci/lambda:build-python3.7' needs to be pulled first.
[2020-05-03T19:13:44Z] DEBUG    - Trying paths: ['/home/circleci/.docker/config.json', '/home/circleci/.dockercfg']
[2020-05-03T19:13:44Z] DEBUG    - No config file found
[2020-05-03T19:13:44Z] DEBUG    - Trying paths: ['/home/circleci/.docker/config.json', '/home/circleci/.dockercfg']
[2020-05-03T19:13:44Z] DEBUG    - No config file found
[2020-05-03T19:13:44Z] DEBUG    - Looking for auth config
[2020-05-03T19:13:44Z] DEBUG    - No auth config in memory - loading from filesystem
[2020-05-03T19:13:44Z] DEBUG    - Trying paths: ['/home/circleci/.docker/config.json', '/home/circleci/.dockercfg']
[2020-05-03T19:13:44Z] DEBUG    - No config file found
[2020-05-03T19:13:44Z] DEBUG    - Looking for auth entry for 'docker.io'
[2020-05-03T19:13:44Z] DEBUG    - No entry found
[2020-05-03T19:13:44Z] DEBUG    - No auth config found
[2020-05-03T19:14:43Z] DEBUG    - Caught downstream error
Traceback (most recent call last):
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/python/codegen.py", line 256, in _docker_build
    user=f"{os.geteuid()}:{os.getgid()}",
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/docker/models/containers.py", line 814, in run
    container, exit_status, command, image, out
docker.errors.ContainerError: Command 'pip install --no-cache-dir --no-color --disable-pip-version-check --upgrade --find-links /project --requirement /project/requirements.txt --target /project/build' in image 'lambci/lambda:build-python3.7' returned non-zero exit status 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/core/cli.py", line 98, in main
    args.command(args)
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/core/submit.py", line 21, in submit
    args.set_default,
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/core/project.py", line 336, in submit
    self._plugin.package(self, zip_file)
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/python/codegen.py", line 180, in package
    self._build(project.root)
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/python/codegen.py", line 199, in _build
    self._docker_build(base_path)
  File "/home/circleci/my-app/.venv/lib/python3.7/site-packages/rpdk/python/codegen.py", line 268, in _docker_build
    raise DownstreamError("Error running docker build") from e
rpdk.core.exceptions.DownstreamError: Error running docker build
../common.mk:50: recipe for target 'sendgrid-whitelabel-domain.zip' failed
make[1]: *** [sendgrid-whitelabel-domain.zip] Error 1
make[1]: Leaving directory '/home/circleci/my-app/sendgrid-whitelabel-domain'
Makefile:9: recipe for target 'sendgrid-whitelabel-domain' failed
make: *** [sendgrid-whitelabel-domain] Error 2

Exited with code exit status 2
```

Note that I use the `setup_remote_docker` circleci directive to make a docker daemon available within the build:

```
Allocating a remote Docker Engine

Assigned Docker Engine request id: 210331
  provisioning: .
Remote Docker engine created. Using VM 'default-69b35ef5-d7e7-4ef4-a1b1-62537571327a'
Created container accessible with:
  DOCKER_CERT_PATH=/tmp/docker-certs299933069
  DOCKER_HOST=tcp://10.32.17.133:2376
  DOCKER_MACHINE_NAME=210331
  DOCKER_TLS_VERIFY=1
  NO_PROXY=127.0.0.1,localhost,circleci-internal-outer-build-agent,10.32.17.133:2376
```

My suspicion is that the issue is that there is a problem with the docker environment that `cfn submit` now creates, which is now 'nested' (though flat I think from the docker daemon perspective.)

I am able to reproduce the failure by creating a container with access to a docker daemon:

```
#cimg/python3.7 is the CircleCI image we're using
docker run -it -v path-to-project:/project -v /var/run/docker.sock:/var/run/docker.sock cimg/python:3.7
sudo apt-get update
sudo apt-get install containerd docker.io
cd /project
make venv # install cfn, etc
sudo bash # docker requires sudo here
. .venv/bin/activate
make
# ... same error as above.
```

When I make the nested container manually:
`sudo docker run -v /var/run/docker.sock:/var/run/docker.sock -it lambci/lambda:build-python3.7 bash`
... I'm able to run the problem pip command successfully
`pip install --no-cache-dir --no-color --disable-pip-version-check --upgrade --find-links
/project --requirement /project/requirements.txt --target /project/build`
... *if* requirements.txt (including 'cloudformation-cli-python-lib==0.0.1') is present, as well as the needed cloudformation-cli-python-lib-0.0.1.tar.gz.

**Comment by mikemccabe** [comment]
I think this discussion in the circleci docs is relevant.

https://circleci.com/docs/2.0/building-docker-images/#mounting-folders

Obviously, not all cfn builds are in circleci, but it would be great if cfn could switch to a file propagation method that worked in that context (as well as outside of it.)

**Comment by wbingli** [comment]
Hey @mikemccabe 

Sorry for the late reply.  The Python plugin `cloudformation-cli-python-plugin` has released 2.0.0 as public PyPI packages. You can follow [the latest guide ](https://github.com/aws-cloudformation/cloudformation-cli-python-plugin/)to build your resource.

In the meanwhile,  you may try to build not using docker by configuring `"use_docker": false` in the `.rpdk-config`.

Let me know if any issue.

Thanks
